torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -0,0 +1,380 @@
+using nss = std::chrono::nanoseconds;
+
+std::atomic<uint32_t> TensorPipeAgent::nextShmAddressID_(0);
+
+TensorPipeAgent::TensorPipeAgent(
+    worker_id_t selfId,
+    std::string selfName,
+    std::shared_ptr<::c10d::Store> addressStore,
+    TensorPipeRpcBackendOptions opts)
+    : RpcAgent(
+          WorkerInfo(std::move(selfName), std::move(selfId)),
+          std::make_unique<RequestCallbackImpl>(),
+          opts.rpcTimeout),
+      context_(tensorpipe::Context::create()),
+      addressStore_(std::move(addressStore)),
+      opts_(std::move(opts)) {
+  // Generate the maps for once.
+  for (const std::pair<std::string, worker_id_t>& kv : opts_.workerNameToId) {
+    const string& workerName = kv.first;
+    worker_id_t workerId = kv.second;
+    workerIdToInfo_.emplace(workerId, WorkerInfo(workerName, workerId));
+    workerNameToInfo_.emplace(workerName, WorkerInfo(workerName, workerId));
+  }
+}
+
+TensorPipeAgent::~TensorPipeAgent() {
+  shutdown();
+}
+
+void TensorPipeAgent::start() {
+  context_->registerTransport(
+      1, "tcp", std::make_shared<tensorpipe::transport::uv::Context>());
+#ifdef TP_ENABLE_SHM
+  context_->registerTransport(
+      0, "shm", std::make_shared<tensorpipe::transport::shm::Context>());
+#endif
+  context_->registerChannelFactory(
+      1,
+      "basic",
+      std::make_shared<tensorpipe::channel::basic::BasicChannelFactory>());
+#ifdef TP_ENABLE_CMA
+  context_->registerChannelFactory(
+      0, "cma", tensorpipe::channel::cma::CmaChannelFactory::create());
+#endif
+
+  std::vector<std::string> addresses = {createUniqueShmAddr(),
+                                        "tcp://127.0.0.1"};
+
+  listener_ = tensorpipe::Listener::create(context_, addresses);
+
+  // Store our own url.
+  const auto address = listener_->url("tcp");
+  const std::vector<uint8_t> selfAddrData(address.begin(), address.end());
+  addressStore_->set(workerInfo_.name_, selfAddrData);
+
+  for (const auto& p : workerNameToInfo_) {
+    const auto& name = p.first;
+    auto nodeAddrData = addressStore_->get(name);
+    auto nodeAddrStr =
+        std::string((const char*)nodeAddrData.data(), nodeAddrData.size());
+    workerNameToURL_.insert({name, nodeAddrStr});
+  }
+
+  listener_->accept([this](
+                        const tensorpipe::Error& error,
+                        std::shared_ptr<tensorpipe::Pipe> pipe) {
+    onListenerAccepted(error, pipe);
+  });
+
+  isRunning_ = true;
+}
+
+void TensorPipeAgent::onListenerAccepted(
+    const tensorpipe::Error& error,
+    std::shared_ptr<tensorpipe::Pipe>& pipe) {
+  // TODO: Handle listener error
+  if (error) {
+    LOG(WARNING) << "got error from listener: " << error.what();
+    return;
+  }
+
+  // Accept the next connection request
+  listener_->accept([this](
+                        const tensorpipe::Error& error,
+                        std::shared_ptr<tensorpipe::Pipe> pipe) {
+    onListenerAccepted(error, pipe);
```

Comment:
The pipe received in this callback is not stored on a member field of the agent, it's just passed around by reference and sometimes captured by copy in a callback. Would it perhaps be more robust to store it, to make sure it's kept alive no matter what?

Reply:
I see your point. The pipe would be kept alive in the next respond() calls. It will be gone when gets an error say closed by the peer. If we would like to keep a copy in agent, it may be overhead to handle the destruction if the connection is closed, or just keep it alive until agent is destructed?